R/R/prophet.R

```diff
@@ -1160,4 +1160,52 @@ plot_yearly <- function(m, uncertainty = TRUE, yearly_start = 0) {
   return(gg.yearly)
 }
 
+#' Return a List having posterior predictive samples
+#'
+#' @param m Prophet model object.
+#' @param future Dataframe that extends forward from the end of m$history for the
+#' requested number of periods.
+#' @return list of sim.values : posterior predictive samples,
+#'                 yhat.mean : mean of predictive samples,
+#'                 lower.value : lower bound of uncertainty interval 
+#'                 upper.value : upper bound of uncertainty interval                
+#' @export
+posterior_analysis <- function(future, m) {
+  
+  df <-  setup_dataframe(m2,future)$df
+  df$trend <- predict_trend(m2, df)
+  
+  n.iterations <- length(m$params$k)
+  samp.per.iter <- max(1, ceiling(m$uncertainty.samples / n.iterations))
+  nsamp <- n.iterations * samp.per.iter  # The actual number of samples
+  
+  seasonal.features <- make_all_seasonality_features(m, df)
+  sim.values <- list("trend" = matrix(, nrow = nrow(df), ncol = nsamp),
+                     "seasonal" = matrix(, nrow = nrow(df), ncol = nsamp),
+                     "yhat" = matrix(, nrow = nrow(df), ncol = nsamp))
+  
+  for (i in 1:n.iterations) {
+    # For each set of parameters from MCMC (or just 1 set for MAP),
+    for (j in 1:samp.per.iter) {
+      # Do a simulation with this set of parameters,
+      sim <- sample_model(m, df, seasonal.features, i)
+      # Store the results
+      for (key in c("trend", "seasonal", "yhat")) {
+        sim.values[[key]][,(i - 1) * samp.per.iter + j] <- sim[[key]]
+      }
+    }
+  }
```

Comment:
There's a fair amount of code duplication between this function and predict_uncertainty, which I'd really prefer to avoid. Let's have a function with all of this code that is used by both this and predict_uncertainty, which function will be called by both. Basically it would be
```
#' Prophet posterior predictive samples.
#'
#' @param m Prophet object.
#' @param df Prediction dataframe.
#'
#' @return List with posterior predictive samples for each component.
#'
sample_posterior_predictive <- function(m, df) {
  # Sample trend, seasonality, and yhat from the extrapolation model.
  n.iterations <- length(m$params$k)
  samp.per.iter <- max(1, ceiling(m$uncertainty.samples / n.iterations))
  nsamp <- n.iterations * samp.per.iter  # The actual number of samples

  seasonal.features <- make_all_seasonality_features(m, df)
  sim.values <- list("trend" = matrix(, nrow = nrow(df), ncol = nsamp),
                     "seasonal" = matrix(, nrow = nrow(df), ncol = nsamp),
                     "yhat" = matrix(, nrow = nrow(df), ncol = nsamp))

  for (i in 1:n.iterations) {
    # For each set of parameters from MCMC (or just 1 set for MAP),
    for (j in 1:samp.per.iter) {
      # Do a simulation with this set of parameters,
      sim <- sample_model(m, df, seasonal.features, i)
      # Store the results
      for (key in c("trend", "seasonal", "yhat")) {
        sim.values[[key]][,(i - 1) * samp.per.iter + j] <- sim[[key]]
      }
    }
  }

#' Prophet uncertainty intervals.
#'
#' @param m Prophet object.
#' @param df Prediction dataframe.
#'
#' @return Dataframe with uncertainty intervals.
#'
predict_uncertainty <- function(m, df) {
  sim.values <- sample_posterior_predictive(m, df)
  # Add uncertainty estimates
  lower.p <- (1 - m$interval.width)/2
  upper.p <- (1 + m$interval.width)/2

  intervals <- cbind(
    t(apply(t(sim.values$yhat), 2, stats::quantile, c(lower.p, upper.p),
            na.rm = TRUE)),
    t(apply(t(sim.values$trend), 2, stats::quantile, c(lower.p, upper.p),
            na.rm = TRUE)),
    t(apply(t(sim.values$seasonal), 2, stats::quantile, c(lower.p, upper.p),
            na.rm = TRUE))
  ) %>% dplyr::as_data_frame()

  colnames(intervals) <- paste(rep(c('yhat', 'trend', 'seasonal'), each=2),
                               c('lower', 'upper'), sep = "_")
  return(intervals)
}

#' Sample from the posterior predictive distribution.
#'
#' @param m Prophet object.
#' @param df Dataframe with dates for predictions (column ds), and capacity
#'  (column cap) if logistic growth.
#'
#' @return A list with items "trend", "seasonal", and "yhat" containing 
#'  posterior predictive samples for that component.
#'
#' @export
predictive_samples <- function(m, df) {
  df <- setup_dataframe(object, df)$df
  sim.values <- sample_posterior_predictive(m, df)
  return(sim.values)
}
```